Emt 2431 - Draft PR

bboothe-branch opened the pull request:
## Reference
EMT 2431 -- <TITLE>.

## Description
<!-- SUFFICIENT DESCRIPTION TO EXPLAIN THE PROBLEM THAT IS BEING SOLVED -->
Added in the main structure of the potential update for Android Billing Library using Reflection

## Testing Instructions
<!-- TESTING INSTRUCTIONS -->
Download and run test app. Currently errors.

## Risk Assessment [`HIGH` || `MEDIUM` || `LOW`]
<!-- CHOOSE ONE OF THE THREE ASSESSMENTS ABOVE -->
<!-- FOR MEDIUM OR HIGH ASSESSMENTS, ADD ADDITIONAL NOTES HERE -->
High Risk as this may break android billing entirely if not done correctly.

- [ ] I, the PR creator, have tested — integration, unit, or otherwise — this code.

## Reviewer Checklist (To be checked off by the reviewer only)

- [ ] JIRA Ticket is referenced in PR title.
- Correctness & Style
    - [ ] Conforms to [AOSP Style Guides](https://source.android.com/setup/contribute/code-style)
    - [ ] Mission critical pieces are documented in code and out of code as needed.
- [ ] Unit Tests reviewed and test issue sufficiently.
- [ ] Functionality was reviewed in QA independently by another engineer on the team.

cc @BranchMetrics/saas-sdk-devs for visibility.

gdeluna-branch commented on Branch-SDK/src/androidTest/AndroidManifest.xml:
```diff
@@ -2,8 +2,8 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android">
 
     <uses-permission android:name="android.permission.INTERNET" />
-    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
```
You can remove commented code

gdeluna-branch commented on Branch-SDK/build.gradle.kts:
```diff
@@ -136,6 +141,18 @@ android {
         isRequired = isReleaseBuild()
     }
+
+    flavorDimensions.add("billing")
```
Let's keep things simple and just check the buildconfig version string and conditionally execute the right library implementation.

bboothe-branch replied:
Got it. Removing flavor dimensions

bboothe-branch commented on Branch-SDK/src/BillingGooglePlayModules/BillingV6/build.gradle:
```diff
@@ -0,0 +1,40 @@
+plugins {
```
You mean having that line in the main gradle file instead? So I'll remove the product flavors.

Doe that replace this? :`compileOnly("com.android.billingclient:billing:6.0.1")`

bboothe-branch commented on settings.gradle.kts:
```diff
@@ -1,5 +1,7 @@
 include(":Branch-SDK")
 include(":Branch-SDK-TestBed")
+include (":BillingGooglePlayModules:BillingV6V7", ":BillingGooglePlayModules:BillingV8")
```
Removed this as well

bboothe-branch commented on Branch-SDK/src/main/java/io/branch/referral/BillingGooglePlayReflection.java:
```diff
@@ -0,0 +1,29 @@
+public class BillingGooglePlayReflection {
+    public static BillingGooglePlayInterface getBillingLibraryVersion() {
+        try {
+            // Check for a class added in version 8.0 or higher
```
There isn't a specific method to grab the version so the checks look for what's available in the version that is being used.

bboothe-branch replied:
Just realizing that was for the reflection section. Replacing it!

gdeluna-branch commented on Branch-SDK/src/main/java/io/branch/referral/BillingGooglePlayReflection.java:
```diff
@@ -0,0 +1,35 @@
+
+        try {
+            int majorIndex = billingClient.indexOf(".");
+            String majorVersion = billingClient.substring(0, majorIndex);
```
Might be worth making this version extraction bit into a utility function since this could be used in other places as well.

bboothe-branch replied:
Created function called dependencyMajorVersionFinder and added it to DependencyUtils.kt

Function takes in a string with the depenency name and pulls out the major version number.

Kept the Version Name extract itself in the BillingGooglePlayReflection class so that the function can take in any method of calling the version name depending on what the class offers.

gdeluna-branch commented on Branch-SDK/src/main/java/io/branch/referral/BillingGooglePlayReflection.java:
```diff
@@ -0,0 +1,35 @@
+            }
+
+        } catch (Exception e) {
+            System.err.println("Error parsing billing client version: " + e.getMessage());
```
Use `BranchLogger.e` instead of `System.err.println`

bboothe-branch replied:
Changed to BranchLogger.e

gdeluna-branch commented on Branch-SDK/build.gradle.kts:
```diff
@@ -13,6 +13,12 @@ jacoco {
     toolVersion = "0.8.10"
 }
 
+//configurations.all {
```
Delete this

bboothe-branch commented:
Hey @gdeluna-branch I updated all of your comments and also added the V8 logic. Is there a better way to reference them in the BillingGooglePlayReflection.java file? Also, I'm hesitant to delete the BillingGooglePlay.kt file until everything is done and working. 

Also still need to build the gradle to reference each version properly.

gdeluna-branch commented:
We'll re-open this after architecture updates.
